File: internal/game/states/tower.go
package states

import (
	"github.com/slycrel/slycrel/internal/game"
	"github.com/slycrel/slycrel/internal/mechanics"
	"github.com/slycrel/slycrel/internal/model"
)

// TowerState - random encounter in the Tower of Wisdom.
// Original: ST_Tower - 4 random outcomes affecting Faith.
type TowerState struct{}

func (TowerState) ID() game.StateID { return "tower" }

func (TowerState) Enter(s *game.Session) {
	s.Character.Location = model.TheTower

	switch mechanics.RandBetween(0, 3) {
	case 0:
		s.IO.Outln("You enter the dark dingy tower, and immediately encounter", true, 3)
		s.IO.Outln("a holy man saying a prayer in some strange language.", true, 3)
		s.IO.Outln("After he finishes, the holy man looks up and smiles,", true, 3)
		s.IO.Outln("happy to see a hero such as yourself. Again, he begins", true, 3)
		s.IO.Outln("to chant out a prayer....", true, 3)
		s.IO.Outln("Thankfully, its a blessing. When he's finished he", true, 3)
		s.IO.Outln("sends you on to do your hero's work.", true, 3)
		s.IO.Cr()
		s.Character.Faith += mechanics.RandBetween(1, 2)
	case 1:
		s.IO.Outln("You enter the tower, hopeful that you will gain some", true, 3)
		s.IO.Outln("much needed wisdom....", true, 3)
		s.IO.Cr()
		s.IO.Outln("Eventually, you encounter a High Priestess. The", true, 3)
		s.IO.Outln("priestess lets out a shriek of laugh, and casts a spell", true, 3)
		s.IO.Outln("on you...", true, 3)
		s.IO.Cr()
		s.IO.Outln("When you awaken, you feel dreadfully weak. When you", true, 3)
		s.IO.Outln("eventually find the strength, you leave the tower as", true, 3)
		s.IO.Outln("quickly as possible.", true, 3)
		s.IO.Cr()
		s.Character.Faith -= mechanics.RandBetween(1, 2)
	case 2:
		s.IO.Outln("As you enter the tower, you notice an evil feeling", true, 3)
		s.IO.Outln("has covered the area. Being a hero, you don't seem to mind.", true, 3)
		s.IO.Outln("After a considerable search, you encounter a wizard. The", true, 3)
		s.IO.Outln("Wizard immediately begins to cast a spell.....", true, 3)
		s.IO.Outln("Suddenly you find yourself back in town,", true, 3)
		s.IO.Outln("with an incredible amount of new energy.", true, 3)
		s.IO.Cr()
		s.Character.Faith += mechanics.RandBetween(1, 3)
	case 3:
		s.IO.Outln("You enter the tower, and a fearful feeling", true, 3)
		s.IO.Outln("immediately consumes your every thought. As you", true, 3)
		s.IO.Outln("peer through the darkness, you see a faint image", true, 3)
		s.IO.Outln("across the room. As you creep closer, you finally", true, 3)
		s.IO.Outln("realize you've encountered a High Priest. Before", true, 3)
		s.IO.Outln("you can even turn around, he casts a spell on you.", true, 3)
		s.IO.Outln("An incredible dizziness immediately overtakes you....", true, 3)
		s.IO.Cr()
		s.IO.Outln("You awaken in the center of town, so weak you can", true, 3)
		s.IO.Outln("barely stand.", true, 3)
		s.IO.Cr()
		s.Character.Faith -= mechanics.RandBetween(1, 2)
	}

	s.IO.PausePrompt("-=Press A Key=-")
	s.SetNextState("town")
}

File: clients/web/game/states/tower.js
import { WhereType } from '../../model/enums.js';
import { randBetween } from '../../mechanics/rand.js';
import { saveCharacter } from '../../store/remote.js';
import { TownState } from './town.js';

// TowerState — port of internal/game/states/tower.go. Four random outcomes
// that adjust Faith up or down.
export class TowerState {
  async enter(session) {
    const c = session.character;
    const { io } = session;
    c.location = WhereType.TheTower;

    switch (randBetween(0, 3)) {
      case 0:
        io.println('You enter the dark dingy tower, and immediately encounter', 3);
        io.println('a holy man saying a prayer in some strange language.', 3);
        io.println('After he finishes, the holy man looks up and smiles,', 3);
        io.println('happy to see a hero such as yourself. Again, he begins', 3);
        io.println('to chant out a prayer....', 3);
        io.println("Thankfully, it's a blessing. When he's finished he", 3);
        io.println("sends you on to do your hero's work.", 3);
        io.cr();
        c.faith += randBetween(1, 2);
        break;
      case 1:
        io.println('You enter the tower, hopeful that you will gain some', 3);
        io.println('much needed wisdom....', 3);
        io.cr();
        io.println('Eventually, you encounter a High Priestess. The', 3);
        io.println('priestess lets out a shriek of laugh, and casts a spell', 3);
        io.println('on you...', 3);
        io.cr();
        io.println('When you awaken, you feel dreadfully weak. When you', 3);
        io.println('eventually find the strength, you leave the tower as', 3);
        io.println('quickly as possible.', 3);
        io.cr();
        c.faith -= randBetween(1, 2);
        break;
      case 2:
        io.println('As you enter the tower, you notice an evil feeling', 3);
        io.println("has covered the area. Being a hero, you don't seem to mind.", 3);
        io.println('After a considerable search, you encounter a wizard. The', 3);
        io.println('Wizard immediately begins to cast a spell.....', 3);
        io.println('Suddenly you find yourself back in town,', 3);
        io.println('with an incredible amount of new energy.', 3);
        io.cr();
        c.faith += randBetween(1, 3);
        break;
      case 3:
        io.println('You enter the tower, and a fearful feeling', 3);
        io.println('immediately consumes your every thought. As you', 3);
        io.println('peer through the darkness, you see a faint image', 3);
        io.println('across the room. As you creep closer, you finally', 3);
        io.println("realize you've encountered a High Priest. Before", 3);
        io.println('you can even turn around, he casts a spell on you.', 3);
        io.println('An incredible dizziness immediately overtakes you....', 3);
        io.cr();
        io.println('You awaken in the center of town, so weak you can', 3);
        io.println('barely stand.', 3);
        io.cr();
        c.faith -= randBetween(1, 2);
        break;
    }

    await saveCharacter(c);
    await io.pausePrompt('-=Press A Key=-');
    session.setNext(new TownState());
  }
}
